File: 1048-longest-string-chain/1048-longest-string-chain.cpp
class Solution {
public:
    int longestStrChain(vector<string>& words) {
        int n = words.size(); 

        auto comparator = [](string a,string b) {
            return a.size() < b.size();
        };

        sort(words.begin(), words.end(), comparator);

        vector<int> dp(n, 1);
        int ans = 0;

        for (int i = 0; i < n; i++){
            for (int j = 0; j < i; j++){
                if (words[j].size() + 1 == words[i].size()){
                    int k = words[j].size(), l = words[i].size();
                    k--;l--;

                    while (k >= 0 && l >= 0){
                        if (words[j][k] == words[i][l]){
                            k--;l--;
                        } else {
                            l--;
                        }
                    }                    

                    if (k < 0){
                        dp[i] = max(dp[i], 1 + dp[j]);
                    }
                }
            }

            ans = max(ans, dp[i]);
        }

        return ans;
    }
};

File: 1048-longest-string-chain/1048-longest-string-chain.java
class Solution {
    public int longestStrChain(String[] words) {
        int n = words.length;
        int[] chainLengths = new int[n];
        Arrays.fill(chainLengths, 1);
        Arrays.sort(words, (a,b) -> Integer.compare(a.length(), b.length()));

        for (int i = 0; i < n; i++) {
            
            for (int j = 0;  j < i; j++) {
                if (isAPredOfB(words[j], words[i])) {
                    chainLengths[i] = Math.max(chainLengths[i], chainLengths[j] + 1);
                }
            }
        }
        
        return Arrays.stream(chainLengths).max().getAsInt();
    }

    

    private boolean isAPredOfB(String a, String b) {
        if (a.length() + 1 != b.length()) return false;
        
        int i = 0, j = 0;
        while (i < a.length() && j < b.length()) {
            if (a.charAt(i) == b.charAt(j)) {
                i++; j++;
            } else {
                j++; // skip one character in b
            }
        }
        return i == a.length(); // all of a must be matched
    }

}